platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/internal/PmdInvoker.java

```diff
@@ -0,0 +1,220 @@
+            // http://sourceforge.net/p/pmd/bugs/1004/
+            // http://java-pmd.30631.n5.nabble.com/pmd-pmd-db05bc-pmd-AntTask-support-for-language-td5710041.html
+            antPmdArgs.put("targetjdk", parameters.getTargetJdk().get().getName());
+
+            htmlFormat = "betterhtml";
+
+            // fallback to basic on pre 5.0 for backwards compatible
+            if (ruleSets.equals(JAVA_BASIC) || ruleSets.equals(ERROR_PRONE)) {
+                ruleSets = BASIC;
+            }
+            if (parameters.getIncrementalAnalysis().get()) {
+                PmdInvoker.assertUnsupportedIncrementalAnalysis(version);
+            }
+        } else if (version.compareTo(VersionNumber.parse("6.0.0")) < 0) {
+            // 5.x
+            if (ruleSets.equals(ERROR_PRONE)) {
+                ruleSets = JAVA_BASIC;
+            }
+            if (parameters.getIncrementalAnalysis().get()) {
+                PmdInvoker.assertUnsupportedIncrementalAnalysis(version);
+            }
+            antPmdArgs.put("threads", parameters.getThreads().get());
+        } else {
+            // 6.+
+            if (parameters.getIncrementalAnalysis().get()) {
+                antPmdArgs.put("cacheLocation", parameters.getIncrementalCacheFile().get().getAsFile());
+            } else {
+                if (version.compareTo(VersionNumber.parse("6.2.0")) >= 0) {
+                    antPmdArgs.put("noCache", true);
+                }
+            }
+            antPmdArgs.put("threads", parameters.getThreads().get());
+        }
+
+        int maxFailures = parameters.getMaxFailures().get();
+        if (maxFailures < 0) {
+            throw new GradleException(String.format("Invalid maxFailures %s. Valid range is >= 0.", maxFailures));
+        }
+
+        String finalHtmlFormat = htmlFormat;
+        List<String> finalRuleSets = ruleSets;
+        List<PmdActionParameters.EnabledReport> reports = parameters.getEnabledReports().get();
+        ant.taskdef(ImmutableMap.of("name", "pmd", "classname", "net.sourceforge.pmd.ant.PMDTask"));
+        invoke(ant, "pmd", antPmdArgs, () -> {
+            parameters.getSource().addToAntBuilder(ant, "fileset", FileCollection.AntType.FileSet);
+            finalRuleSets.forEach(rule -> {
+                invoke(ant, "ruleset", rule);
+            });
+            parameters.getRuleSetConfigFiles().forEach(ruleSetConfig -> {
+                invoke(ant, "ruleset", ruleSetConfig);
+            });
+
+            FileCollection auxClasspath = parameters.getAuxClasspath().filter(new FileExistFilter());
+            if (!auxClasspath.isEmpty()) {
+                auxClasspath.addToAntBuilder(ant, "auxclasspath", FileCollection.AntType.ResourceCollection);
+            }
+
+            reports.forEach(report -> {
+                File file = report.getOutputLocation().getAsFile().get();
+                assert file.getParentFile().exists();
```

Comment:
This needs to be upgraded to a proper check, as this would only run under `-ea`, which I don't think we use.

The previous Groovy assertions were always enabled.